src/browser/fake_network.js

```diff
@@ -43,6 +58,220 @@ function iptolong(parts) {
+        this.ipv4_payload_view = new DataView(eth_frame.buffer, offset + IPV4_PAYLOAD_OFFSET, IPV4_PAYLOAD_SIZE);
+        this.udp_payload_view = new DataView(eth_frame.buffer, offset + UDP_PAYLOAD_OFFSET, UDP_PAYLOAD_SIZE);
+
+        for(const view of [this.eth_frame_view, this.eth_payload_view, this.ipv4_payload_view, this.udp_payload_view]) {
+            view.setArray = this.view_setArray.bind(this, view);
+            view.setString = this.view_setString.bind(this, view);
+            view.setInetChecksum = this.view_setInetChecksum.bind(this, view);
+        }
+
+        this.text_encoder = new TextEncoder();
+    }
+
+    /**
+     * Copy given data array into dst_view[dst_offset], return number of bytes written.
+     *
+     * @param {DataView} dst_view
+     * @param {number} dst_offset
+     * @param data
+     */
+    view_setArray(dst_view, dst_offset, data)
+    {
+        this.eth_frame.set(data, dst_view.byteOffset + dst_offset);
+        return data.length;
+    }
+
+    /**
+     * UTF8-encode given string into dst_view[dst_offset], return number of bytes written.
+     *
+     * @param {DataView} dst_view
+     * @param {number} dst_offset
+     * @param {string} str
+     */
+    view_setString(dst_view, dst_offset, str)
+    {
+        const ofs = dst_view.byteOffset + dst_offset;
+        const result = this.text_encoder.encodeInto(str, ofs ? this.eth_frame.subarray(ofs) : this.eth_frame);
+        return result.written;
+    }
+
+    /**
+     * Calculate 16-bit internet checksum for dst_view[0 : length], store result in dst_view[dst_offset].
+     * Source: RFC768 and RFC1071 (chapter 4.1).
+     *
+     * @param {DataView} dst_view
+     * @param {number} dst_offset
+     * @param {number} length
+     * @param {number} checksum
+     */
+    view_setInetChecksum(dst_view, dst_offset, length, checksum)
+    {
+        const data = this.eth_frame;
+        const offset = dst_view.byteOffset;
+        const uint16_end = offset + (length & ~1);
+        for(let i = offset; i < uint16_end; i += 2) {
+            checksum += data[i] << 8 | data[i+1];
+        }
+        if(length & 1) {
+            checksum += data[uint16_end] << 8;
+        }
+        while(checksum >> 16) {
+            checksum = (checksum & 0xffff) + (checksum >> 16);
+        }
+        dst_view.setUint16(dst_offset, ~checksum);
+    }
+
+    /**
+     * Encode and return ethernet packet for given spec.
+     * TODO: what about the trailing 32-bit ethernet checksum?
+     *
+     * @param {Object} spec
+     */
+    encode_packet(spec)
+    {
+        dbg_assert(spec.eth);
+        this.eth_frame.fill(0);
+        const length = write_eth(spec, this.eth_frame_view, this);
+        return this.eth_frame.subarray(0, length);
+    }
+}
+
+const ethernet_encoder = new EthernetPacketEncoder();
+
+function make_packet(spec)
+{
+    return ethernet_encoder.encode_packet(spec);
+}
```

Comment:
I'm really not a fan of the `EthernetPacketEncoder` class. Is there any way we can write

```javascript
function make_packet(spec, out)
{
     // encode spec into out, which is an ArrayBuffer(View)
     // pure function
}
```

And have the caller cache `out`?

We could also consider changing `receive` to take a high-level packet description and tcp/udp payload, and construct the packet in place (in ne2k/virtio), which would be even fewer copies.

Reply:
I removed EthernetPacketEncoder, it was never meant to stay.

The current state is somewhere in the middle ground towards changing `receive` to take a package descriptor object containing a reference to a GrowableRingbuffer (TCP) or ArrayBuffer (UDP). I'll do this next.

More details in the patch description.